### PR TITLE
Update PETSc to 3.20.3, update libmesh

### LINKED ISSUE #26461: Update PETSc to 3.20.3
## Reason
PETSc has released 3.20.3 as of 2023-12-30, and the PETSc submodule in MOOSE should keep up with the newest features and bugfixes.

## Design
Update the MOOSE submodule to the PETSc 3.20.3 commit.

## Impact
New PETSc capability, improvements, and bugfixes.

### PR DESCRIPTION
**Pull request by cticenhour**
Closes #26461

### CODE REVIEW

**Comment by cticenhour** [comment]
@roystgnr I'd like to get one round of testing in on this before piling more on, but would you be interested in a libMesh update going in alongside this?

**Comment by roystgnr** [comment]
> would you be interested in a libMesh update going in alongside this?

Probably not.  I've got a minor fix that would be useful in MOOSE, but not worth the hassle alone, and I've got a major branch that isn't likely to be ready to merge before the end of the week.

**Comment by cticenhour** [comment]
Alright - will ping again if this ends up lingering through Friday or Monday.

**Comment by lindsayad** [comment]
I would be interested in a libmesh update

**Comment by lindsayad** [comment]
Can't build https://github.com/idaholab/moose/pull/26477 without a libmesh update. A couple cool new things in libMesh directly leveraged in that MOOSE PR are `L2_HIERARCHIC_VEC` and the ability to form a callback post linear-solve, right before the line search.

That being said, I would be willing to wait for @roystgnr's major branch to merge (is it libmesh/libmesh#3759 or something else?)

**Comment by roystgnr** [comment]
I've pushed newsletter docs to https://github.com/roystgnr/moose/commit/1bfae7deda9d0f2cc451ba84a81dd5ab1615c393 if you want to cherry-pick that to go alongside a libmesh submodule update.

**Comment by cticenhour** [comment]
Thanks for your input! I will update this PR tomorrow to include a libMesh update as well. The only thing holding this back on testing at the moment is a FBLASLAPACK download issue during PETSc alt container creation....working with Logan on it.

File: modules/doc/content/newsletter/2024/2024_01.md
Comment (cticenhour) [inline modules/doc/content/newsletter/2024/2024_01.md:12]:
```suggestion
### `2024.01.05` Update
```

**Comment by roystgnr** [comment]
If there's more futzing to be done anyway, it wouldn't hurt to get libMesh up to https://github.com/libMesh/libmesh/commit/3fcc7e7099f1b4abee4a048fee27a1f01f5e8b18 - the additional changelog entries would be:

- Fixes for VTK .pvtu reading, and preservation of libMesh node, element, and subdomain ids written to VTK files
- `Elem::volume()` no longer can throw an exception when called on an element without an invertible mapping; hence `print_info()` and `get_info()` are now safe to use when emitting debugging messages about twisted geometric elements.

IMHO neither of those are super high priority if you just want to focus on getting the PETSc issues fixed and getting this in.

**Comment by cticenhour** [comment]
@roystgnr The GCC min / PETSc-alt failures are all tied to the FBLASLAPACK download issue (which is preventing the build of a new container for those tests), but the new app failures I am seeing are due to the libMesh additions. Do you mind taking a look today? They all seem to be related to the nodeset changes or similar.

**Comment by roystgnr** [comment]
Will do!  Looks like we've got a segfault in a Malamute test and 7 (looks like 2 independent) exodiffs in Bison tests?

**Comment by roystgnr** [comment]
The good news is that I understand the MALAMUTE failure now.

The bad news is that the MALAMUTE failure is coming from MOOSE doing something unsupported and, up until now, getting away with it via dumb luck.

That input deck creates a mesh of `QUAD8` elements.  It then has MOOSE create Edge3 elements along one side of some of them, because `formulation = mortar`, and MOOSE manually does a `edge3->set_interior_parent(quad8)` on each.  It then has MOOSE convert all the Quad8 elements to Quad9, because `second_order = true`, and the trouble with this of course is that *you can't actually convert one class into another in C++*.  What libMesh does to "convert" a mesh like this is to delete all the lower-order elements and allocate higher-order elements to replace them.  This has apparently (probably?) been working out for MALAMUTE in the past because the allocated elements have been ending up at the same addresses in memory!?  And then those `interior_parent` pointers stay valid.  But nothing guarantees that happening, and any kind of modification of the `all_second_order` code (much less my recent extensive modifications) can stop it from happening. (as could a change of compiler, or of compiler flags, or of day of the week...)

I'm trying to figure out a way to make this robust.

**Comment by lindsayad** [comment]
Where is MOOSE doing this? Doesn’t moose rely on libMesh’s api for converting the mesh to second order?

**Comment by lindsayad** [comment]
Haven’t you been doing a bunch of work in these mesh order upgrade methods? Is it possible the behavior is changing in this update? It would seem like quite a coincidence for this to have been working for all CI runs for months

**Comment by roystgnr** [comment]
> Where is MOOSE doing this? Doesn’t moose rely on libMesh’s api for converting the mesh to second order?

The `set_interior_parent()` calls are in `LowerDBlockFromSidesetGenerator`, I believe.  The `all_second_order()` call is in `SetupMeshAction`.  It is using the libMesh API, but the problem is that even the libMesh API can't guarantee that deallocation and reallocation of objects gives you back the same pointers if only you do it in just the right order.  Maybe there's a way to do it with fancy C++ tricks, if we destruct the existing `Elem` without freeing its memory and then you use placement-`new` to put the new `Elem` in its place (assuming all `Elem` subclasses are exactly the same size, which maybe?), but we don't do that.  We just `delete` the existing `Elem` and `new` a replacement, and the memory allocator has apparently been returning the same pointer, but it's not at all guaranteed to.

> Haven’t you been doing a bunch of work in these mesh order upgrade methods? Is it possible the behavior is changing in this update? It would seem like quite a coincidence for this to have been working for all CI runs for months

It's definitely *triggered* by this update (something is performing another allocation in between the `delete` and the `new`? I'm now doing the `new` before the `delete`?), but it's not *caused* by this update.  I like the "nasal demons" joke about Undefined Behavior, where a compiler is allowed to respond to UB by making demons come out your nose, but really the most insidious allowed response to UB is the one we're seeing here: it's allowed to work the way you hoped right up until for no apparent reason it stops working.

On the bright side, the refactoring should at least make it possible to fix every case of this all in one place.

On the not-so-bright side, the fix is going to make some of my new capabilities asymptotically slower.  We can't safely upgrade O(N_range) elements until after we first check O(N_elem) elements to see if they have any `interior_parent()` links that need to be moved as part of the upgrade.  I don't think anybody's going to need them to be `O(N_range)` but it would have been nice to have had the *option*.

**Comment by cticenhour** [comment]
@roystgnr FYI - libMesh is now up to https://github.com/libMesh/libmesh/commit/3fcc7e7099f1b4abee4a048fee27a1f01f5e8b18 and I've rebased on current devel.

**Comment by roystgnr** [comment]
The Bison failure is independent of the Malamute failure, so this libMesh upgrade had *two* test-breaking bugs.  (insert Simpsons meme here: "two test-breaking bugs *so far*")

It's definitely https://github.com/libMesh/libmesh/pull/3752 triggering the Bison regressions, but I've only figured that out because they're easy to reproduce and bisect, not because I've figured out a cause.

Good news is that the outputs are barely regressions.  Here's the part of the diff being complained about:
```
<  node_ns1 = 1, 2, 5, 6, 9, 11, 13, 15, 16, 19, 20, 23, 24, 27, 29, 31 ;
---
>  node_ns1 = 1, 2, 5, 6, 9, 9, 11, 11, 13, 13, 15, 16, 19, 20, 23, 24, 27, 29, 31 ;
```
The careful observer will note that these nodesets are actually identical as *sets*, but apparently both ExodusII and libMesh internals are happy to write redundancies into a vector representing a set, and even Paraview is happy to retain the redundancies, and fortunately exodiff is paranoid enough to yell at us about them.

Bad news is that, although I'm still trying to track down where the redundancies are coming from, my best guess right now is that we've got dangling pointers to the discarded copies of nodes merged via a mesh stitching, in which case "barely regressions" in the files are covering up "more horrible Undefined Behavior waiting to wreck us" in RAM.

**Comment by roystgnr** [comment]
It's not horrible UB this time.  The same node is getting added with the same id twice to the multimap we use.  I really need to change the data structure there; we do have an awful `get_nodeset_map()` raw accessor so it'd technically be an API change, but MOOSE is only using it via a reference-to-auto and hopefully apps are being equally hands-off...

But in the meantime, it looks like I may have just flubbed something in my changes to `BoundaryInfo::clear_stitched_boundary_side_ids()`.  Checking now.

**Comment by roystgnr** [comment]
Damn, even while sick I should have known to check on this.  Those SCHEMADIFF bugs are from the new field in our VTK output.  I'll see about regolding and verifying and pushing updates.

**Comment by cticenhour** [comment]
@roystgnr Just took a peek at the current in-progress testing - framework is now in good shape, but one more test to fix in modules related to IGA (`tensor_mechanics/examples/cframe_iga.cframe`).

**Comment by roystgnr** [comment]
Well, that was a terrifying red herring.  I reran the example, copied the files to gold/, looked at `git diff` before committing, saw a *huge* diff in the values section, panicked over which of my IGA bugfixes over the past few months could possibly have required a value regolding but not already *prompted* a value regolding, dove into git logs ... then took another look at the `git diff`, which showed a bunch of stuff changing in like the 17th digit, ordinary FP error.  I copied *just* the `libmesh_node_id` array into the old gold, reran the example, and it passed the more-floating-point-friendly VTK XMLDiff just fine.

File: libmesh
Comment (roystgnr) [reply libmesh:1]: Hell yeah it does

**Review by roystgnr — APPROVED** [review]
I can't speak to the PETSc part of the change, but everything on the libMesh side should be ready to go.

**Review by loganharbour — CHANGES_REQUESTED** [review]
Going to add in a minor apptainer change and one more libmesh bump.

**Comment by roystgnr** [comment]
> everything on the libMesh side should be ready to go.

I'd like to thank Nemesis for only striking me down with a new bug report from Alex, and not with the lightning bolt that my hubris above may have deserved.

I'm looking into it now.

File: scripts/tests/versioner_hashes.yaml
Comment (loganharbour) [reply scripts/tests/versioner_hashes.yaml]: Sad, silly mistake

**Comment by cticenhour** [comment]
@milljm Could you please give us another set of eyes for these changes? Thanks!

**Review by milljm — APPROVED** [review]
For the Conda stuff, all looks well!

**Comment by roystgnr** [comment]
The fix for that last remaining libMesh bug is in https://github.com/libMesh/libmesh/pull/3772

After we get that to master (hopefully tomorrow), we can update the submodule here and we'll have taken care of everything outstanding I know about.

I'm thinking my approval for merging as-is still stands, though, if anyone is chomping at the bit.  The updates in this PR right now include fixes for potentially more serious and definitely more wide-ranging bugs, in long-standing use cases where as far as I can tell we've so far been saved from data corruption by kindly compilers and dumb luck, whereas the bug fixed in https://github.com/libMesh/libmesh/pull/3772 only affects Tet14 refinement, a relatively new feature.

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
so we didn't get Roy's last fix in?

**Comment by roystgnr** [comment]
Nope; sorry.  Hope you're okay working from a manual submodule update until next month.

**Comment by roystgnr** [comment]
Looks like we missed the `LIBMESH_REV` update in `docker_ci/Dockerfile` here?  And yet that didn't cause any problems for users?

**Comment by loganharbour** [comment]
> Looks like we missed the `LIBMESH_REV` update in `docker_ci/Dockerfile` here? And yet that didn't cause any problems for users?

I need to remove the `docker_ci` folder; it's not used anymore.

**Pull request merged**
